clients/include/utility.hpp

```diff
@@ -968,6 +969,269 @@ inline void host_csr_to_csc(int                     M,
     csc_col_ptr[0] = base;
 }
 
+template <typename T>
+inline void host_csr_to_bsr(hipsparseDirection_t    direction,
+                            int                     M,
+                            int                     N,
+                            int                     block_dim,
+                            int&                    nnzb,
+                            hipsparseIndexBase_t    csr_base,
+                            const std::vector<int>& csr_row_ptr,
+                            const std::vector<int>& csr_col_ind,
+                            const std::vector<T>&   csr_val,
+                            hipsparseIndexBase_t    bsr_base,
+                            std::vector<int>&       bsr_row_ptr,
+                            std::vector<int>&       bsr_col_ind,
+                            std::vector<T>&         bsr_val)
+{
+    int mb = (M + block_dim - 1) / block_dim;
+    int nb = (N + block_dim - 1) / block_dim;
+
+    // quick return if block_dim == 1
+    if(block_dim == 1)
+    {
+        bsr_row_ptr.resize(mb + 1, 0);
+#ifdef _OPENMP
+#pragma omp parallel for schedule(dynamic, 1024)
+#endif
+        for(size_t i = 0; i < csr_row_ptr.size(); i++)
+        {
+            bsr_row_ptr[i] = (csr_row_ptr[i] - csr_base) + bsr_base;
+        }
+
+        nnzb = bsr_row_ptr[mb] - bsr_row_ptr[0];
+
+        bsr_col_ind.resize(nnzb, 0);
+        bsr_val.resize(nnzb * block_dim * block_dim, static_cast<T>(0));
+#ifdef _OPENMP
+#pragma omp parallel for schedule(dynamic, 1024)
+#endif
+        for(size_t i = 0; i < csr_col_ind.size(); i++)
+        {
+            bsr_col_ind[i] = (csr_col_ind[i] - csr_base) + bsr_base;
+        }
+#ifdef _OPENMP
+#pragma omp parallel for schedule(dynamic, 1024)
+#endif
+        for(size_t i = 0; i < csr_val.size(); i++)
+        {
+            bsr_val[i] = csr_val[i];
+        }
+
+        return;
+    }
+
+    // determine number of non-zero block columns for each block row of the bsr matrix
+    bsr_row_ptr.resize(mb + 1, 0);
+
+    bsr_row_ptr[0] = bsr_base;
+
+#ifdef _OPENMP
+#pragma omp parallel for schedule(dynamic, 1024)
+#endif
+    for(int i = 0; i < mb; i++)
+    {
+        int start = csr_row_ptr[i * block_dim] - csr_base;
+        int end   = csr_row_ptr[std::min(M, block_dim * i + block_dim)] - csr_base;
+
+        std::vector<int> temp(nb, 0);
+        for(int j = start; j < end; j++)
+        {
+            int blockCol = (csr_col_ind[j] - csr_base) / block_dim;
+            temp[blockCol]         = 1;
+        }
+
+        int sum = 0;
+        for(int j = 0; j < temp.size(); j++)
+        {
+            sum += temp[j];
+        }
+
+        bsr_row_ptr[i + 1] = sum;
+    }
+
+    for(int i = 0; i < mb; i++)
+    {
+        bsr_row_ptr[i + 1] += bsr_row_ptr[i];
+    }
+
+    nnzb = bsr_row_ptr[mb] - bsr_row_ptr[0];
+
+    // find bsr col indices array
+    bsr_col_ind.resize(nnzb, 0);
+    bsr_val.resize(nnzb * block_dim * block_dim, static_cast<T>(0));
```

Comment:
need to change `static_cast<T>(0)` with `make_DataType<T>(0)` else nvcc will fail